src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs

```diff
@@ -81,40 +87,63 @@ public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context,
+                if (dynamicControllerMetadata != null)
+                {
+                    dynamicValues = dynamicControllerMetadata.Values;
+                }
+                else if (transformerMetadata != null)
+                {
+                    var transformer = (DynamicRouteValueTransformer)httpContext.RequestServices.GetRequiredService(transformerMetadata.SelectorType);
+                    dynamicValues = await transformer.TransformAsync(httpContext, originalValues);
+                }
+                else
+                {
+                    // Not a dynamic controller.
+                    continue;
+                }
 
-                var metadata = endpoint.Metadata.GetMetadata<DynamicControllerMetadata>();
-                if (metadata == null)
+                if (dynamicValues == null)
                 {
+                    candidates.ReplaceEndpoint(i, null, null);
                     continue;
                 }
 
-                var matchedValues = candidates[i].Values;
-                var endpoints = _selector.SelectEndpoints(metadata.Values);
-                if (endpoints.Count == 0)
+                var endpoints = _selector.SelectEndpoints(dynamicValues);
+                if (endpoints.Count == 0 && dynamicControllerMetadata != null)
                 {
-                    // If there's no match this is a configuration error. We can't really check
+                    // If there's no match for a fallback is a configuration error. We can't really check
                     // during startup that the action you configured exists.
                     throw new InvalidOperationException(
                         "Cannot find the fallback endpoint specified by route values: " + 
-                        "{ " + string.Join(", ", metadata.Values.Select(kvp => $"{kvp.Key}: {kvp.Value}")) + " }.");
+                        "{ " + string.Join(", ", dynamicValues.Select(kvp => $"{kvp.Key}: {kvp.Value}")) + " }.");
+                }
+                else if (endpoints.Count == 0)
+                {
+                    candidates.ReplaceEndpoint(i, null, null);
+                    continue;
                 }
 
-                var replacement = endpoints[0];
-                
                 // We need to provide the route values associated with this endpoint, so that features
                 // like URL generation work.
-                var values = new RouteValueDictionary(metadata.Values);
+                var values = new RouteValueDictionary(dynamicValues);
 
                 // Include values that were matched by the fallback route.
-                foreach (var kvp in matchedValues)
+                foreach (var kvp in originalValues)
                 {
                     values.TryAdd(kvp.Key, kvp.Value);
                 }
 
+                var replacement = _selector.SelectBestEndpoint(httpContext, values, endpoints);
```

Comment:
This is where we call into the action constraints. 

The reasons that I went this way instead of using something new to solve the problem:
- This is the simplest thing to do for now
- I'm not sure how common this requirement is outside of MVC.

The thing that this solves is selecting between a GET endpoint and a POST endpoint, etc. It seems obvious someone layering a CMS on top of MVC needs this functionality. Maybe I am just being stupid and this needs to be *built-in*.

Reply:
This is the part I don't like. I don't care that it is slow, but I think having an endpoint routing feature (new hotness) reuse action constraints (old and busted) is going to become more and more troublesome. There are already matcher policies in endpoint routing - host matching - that have no equivalent in action constraints.

Reply:
My memory is hazy about what this PR use to have to what is has now.

What did you do to remove this?

Reply:
We did a much more complex (and better) thing. We use `IEndpointSelectorPolicy` to disambiguate instead of `IActionConstraint`. This means that all of our built-in policies implement both the DFA mode and selector policy mode (already done).